crypto/crypto.lurk

```diff
@@ -0,0 +1,96 @@
+; crypto.lurk
+; 
+; simple crypto-related functions
+;
+; load in repl as:
+; > :load "crypto.lurk"
+
+
+(letrec (
+
+    ; parity tests
+    (odd (lambda (a) (< (/ a 2) 0)))
+    (even (lambda (a) (eq NIL (odd a))))
+
+    ; square & multiply exponentiation
+    (fastexp (lambda (b e)
+        (if (= e 0) 1
+            (if (even e) (fastexp (* b b) (/ e 2))
+                (* b (fastexp (* b b) (/ (- e 1) 2)))))))
+
+    ; GCD
+    (gcd (lambda (a b) 
+        (if (= a b) a
+            (if (> a b) (gcd (- a b) b)
+                (gcd (- b a) a)))))
+
+    ; hash function, poseidon-like
+    ; 2-element permutation used by the hash
+    ; NOT cryptographically safe
+    (permute (lambda (a b)
+        (letrec ( 
+            (rnd (lambda (i a b) (let (
+                (mat1 (lambda (a b) (+ (/ a 3) (* b 5))))
+                (mat2 (lambda (a b) (+ (/ a 7) (* b 9))))
+                (sbox (lambda (a) (fastexp a 5)))
```

Comment:
Cool. Obviously in this case, it would be cheaper just to expand by hand. I like the explanatory library function, just thought I would mention it as an aside.